File: designsystem/src/androidMain/kotlin/me/bookk/designsystem/components/MultiPicker.kt
package me.bookk.designsystem.components

import androidx.compose.animation.animateContentSize
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import dev.icerock.moko.resources.compose.localized
import me.bookk.designsystem.theme.typography.secondary
import me.bookk.designsystem.uistate.MultiPickerState
import me.bookk.designsystem.uistate.PickerPresentation

@Composable
fun <T : PickerPresentation> MultiPicker(
    state: MultiPickerState<T>,
    modifier: Modifier = Modifier,
    pickerContent: @Composable () -> Unit,
    itemContent: @Composable (T, onItemRemove: () -> Unit) -> Unit
) {
    Column(modifier.fillMaxWidth()) {
        Header(state.pickerTitle.localized())
        AppCard(Modifier.fillMaxWidth()) {
            Column(Modifier.animateContentSize()) {
                val placeholder = state.placeholder
                if (state.selectedItems.isEmpty() && placeholder != null) {
                    Text(
                        placeholder.localized(),
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(vertical = 16.dp),
                        textAlign = TextAlign.Center,
                        style = MaterialTheme.typography.bodyMedium.secondary()
                    )
                }
                state.selectedItems.forEach {
                    itemContent(it) {
                        state.onItemsRemoveRequested(listOf(it))
                    }
                }
                if (state.isEditable) {
                    AlignStartTextButton(
                        state.addItemButton,
                        modifier = Modifier.fillMaxWidth()
                    ) {
                        state.isPickerVisible = true
                    }
                }
            }
        }
    }
    if (state.isPickerVisible) {
        pickerContent()
    }
}

File: iosApp/iosApp/DesignSystem/Picker/MultiPicker/MultiPicker.swift
import SwiftUI
import shared

struct MultiPicker<ItemContent: View, SheetContent: View>: View {
    @Bindable private var state: IOSMultiPickerState

    private let pickerContent: () -> SheetContent
    private let itemContent: (PickerPresentation, @escaping () -> Void) -> ItemContent

    init(
        _ state: MultiPickerState,
        @ViewBuilder pickerContent: @escaping () -> SheetContent,
        @ViewBuilder itemContent: @escaping (PickerPresentation, @escaping () -> Void) -> ItemContent
    ) {
        self._state = Bindable(wrappedValue: IOSMultiPickerState.cast(state))
        self.pickerContent = pickerContent
        self.itemContent = itemContent
    }

    var body: some View {
		if state.selectedItems.isEmpty, let placeholder = state.placeholder {
			Text(placeholder.localized())
				.font(.body)
				.foregroundStyle(AppColors.secondary)
				.multilineTextAlignment(.center)
				.frame(maxWidth: .infinity)
				.padding(.vertical, 16)
		}
		ForEach(state.selectedItems, id: \.pickerItemId) { item in
			itemContent(item) {
				withAnimation {
					state.onItemsRemoveRequested([item])
				}
			}
		}
		if state.isEditable {
			TextButton(state.addItemButton) {
				state.isPickerVisible = true
			}
			.buttonStyle(.textInList)
			.sheet(isPresented: Binding(
				get: { state.isPickerVisible },
				set: { state.isPickerVisible = $0 }
			)) {
				pickerContent()
			}
		}
    }
}
